### PR TITLE
add metrics API proposal

### PR DESCRIPTION
**Pull request by MiroDudik**
addresses #3

### CODE REVIEW

File: api/METRICS.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
+# API proposal for metrics
+
+## Example
+
+```python
+# For most sklearn metrics, we would have their group version that returns a Bunch with fields
+#   * overall: overall metric value
+#   * by_group: a dictionary that maps sensitive feature values to metric values
+
+summary = accuracy_score_by_group(y_true, y_pred, sensitive_features=sf, **other_kwargs)
```
Comment (riedgar-ms) [inline api/METRICS.md]: So, basically delete all the 'helper' fields from the `GroupMetricResult` and turn it into a `Bunch`. If we want to eliminate types where possible, this seems reasonable.

File: api/METRICS.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+```python
+# For most sklearn metrics, we would have their group version that returns a Bunch with fields
+#   * overall: overall metric value
+#   * by_group: a dictionary that maps sensitive feature values to metric values
+
+summary = accuracy_score_by_group(y_true, y_pred, sensitive_features=sf, **other_kwargs)
+
+# Exporting into pd.Series or pd.DataFrame in not too complicated
+
+series = pd.Series({**summary.by_group, 'overall': summary.overall})
```
Comment (riedgar-ms) [inline api/METRICS.md:16]: Presumably there's an issue here if one of the groups is called 'overall' but we put that down as a user error?

File: api/METRICS.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+acc_ratio = accuracy_score_ratio(y_true, y_pred, sensitive_features=sf, **other_kwargs)
+acc_group_min = accuracy_score_group_min(y_true, y_pred, sensitive_features=sf, **other_kwargs)
+```
+
+## Functions
+
+*Function signatures*
+
+```python
+metric_by_group(metric, y_true, y_pred, *, sensitive_features, **other_kwargs)
```
Comment (riedgar-ms) [inline api/METRICS.md]: Where `metric` has signature `metric(y_true, y_pred, **kwargs)` right? Also, I don't think we need the single `*` in the signature (and it might even be incompatible with **kwargs on the end)

File: api/METRICS.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+# return the summary for the provided metrics
+
+make_metric_by_group(metric)
```
Comment (riedgar-ms) [inline api/METRICS.md]: Should we add something to the name to make it clear that it's returning another function?

File: api/METRICS.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+# return a callable object <metric>_by_group:
+# <metric>_by_group(...) = metric_by_group(<metric>, ...)
+
+# Transformation functions returning scalars
+difference_from_summary(summary)
```
Comment (riedgar-ms) [inline api/METRICS.md:52]: How do all of these behave when presented with non-scalar metrics?

File: api/METRICS.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ratio_from_summary(summary)
+group_min_from_summary(summary)
+group_max_from_summary(summary)
+
+# Metric-specific functions returing summary and scalars
+<metric>_by_group(y_true, y_pred, *, sensitive_features, **other_kwargs)
```
Comment (riedgar-ms) [inline api/METRICS.md]: Do you mean that we provide `accuracy_by_group`, `recall_by_group`, `precision_by_group` etc. as we do already, and then also `accuracy_difference`, `recall_difference`, `precision_difference` etc.? If so, we probably want to invest in some code generation, rather than handwriting them.

File: api/METRICS.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+|transformation function|output|metric-specific function|code|aif360|
+|-----------------------|------|------------------------|----|------|
+|`difference_from_summary`|max - min|`<metric>_difference`|D|unprivileged - privileged|
+|`ratio_from_summary`|min / max|`<metric>_ratio`|R| unprivileged / privileged|
+|`group_min_from_summary`|min|`<metric>_group_min`|Min| N/A |
+|`group_max_from_summary`|max|`<metric>_group_max`|Max| N/A |
+
+*Supported metric-specific functions*
+
+|metric|variants|task|notes|aif360|
```
Comment (riedgar-ms) [inline api/METRICS.md:87]: 'Variant' references the 'code' in the table above? With 'G' meaning the 'basic group metric' ?

File: api/METRICS.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+|`accuracy_score`| G,D,R,Min | class | sklearn | `accuracy` |
+|`balanced_accuracy_score` | G | class | sklearn | - |
+|`mean_absolute_error` | G,D,R,Max | class,reg | sklearn | class only: `error_rate`
+|`false_positive_rate` | G,D,R | class | | &#x2713; |
+|`false_negative_rate` | G | class | | &#x2713; |
+|`true_positive_rate` | G,D,R | class | | &#x2713; |
+|`true_negative_rate` | G | class | | &#x2713; |
+|`equalized_odds` | D,R | class | max of difference or ratio under `true_positive_rate`, `false_positive_rate` | - |
+|`precision_score`| G | class | sklearn | &#x2713; |
+|`recall_score`| G | class | sklearn | &#x2713; |
```
Comment (riedgar-ms) [inline api/METRICS.md:101]: According to wikipedia, 'recall' *is* the 'true positive rate' .... which is listed above and with different variants. Shouldn't the variants be the same, and should we have all alternative names for all of TPR, FNR etc. ?

Comment (MiroDudik) [reply api/METRICS.md:101]: I'd like to include all variants `{true,false}_positive_rate`, because they are used to define the "equal opportunity" and "equalized odds" disparities. The reason for including `false_negative_rate` is that it is a natural error metric. The reason for `true_negative_rate` is just symmetry (it would be weird if it was missing). The reason for including recall, but not other aliases is that `recall_score` is already a standard `sklearn` metric.

File: api/METRICS.md (hunk 4 above)
Comment (MiroDudik) [reply api/METRICS.md]: Do you have an alternative name in mind? The current proposal is analogous to [sklearn.metrics.make_scorer](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html)

File: api/METRICS.md (hunk 3 above)
Comment (MiroDudik) [reply api/METRICS.md]: Correct re. signature of `metric`. Elsewhere we use `sensitive_features` as the keyword-only argument, so I'd like to keep the same convention.

File: api/METRICS.md (hunk 5 above)
Comment (MiroDudik) [reply api/METRICS.md:52]: I don't think we necessarily need to have a well-defined behavior for non-scalars. For `numpy.ndarray`, we could go with the pointwise operations.

File: api/METRICS.md (hunk 6 above)
Comment (MiroDudik) [reply api/METRICS.md]: See the table below--we probably won't implement that many metric-specific transformations, so perhaps this is not such a major issue. But open to automating.

File: api/METRICS.md (hunk 7 above)
Comment (MiroDudik) [reply api/METRICS.md:87]: Yes. I'll make this clearer in the next commit.

File: api/METRICS.md (hunk 2 above)
Comment (MiroDudik) [reply api/METRICS.md:16]: Correct--we let the user handle any exporting and deal with naming conflicts.

File: api/METRICS.md (hunk 8 above)
Comment (romanlutz) [reply api/METRICS.md:101]: I'm also curious about the alternative names. Equal opportunity is another one that might pop up

Comment (MiroDudik) [reply api/METRICS.md:101]: I prefer starting leaner, so I'd much rather cover the alternative names in the documentation for the time being. If this becomes an issue, we can add additional aliases later.

File: api/METRICS.md (hunk 4 above)
Comment (riedgar-ms) [reply api/METRICS.md]: But it's not making a `metric_by_group()` surely? What it's returning is a function which can be called with y_true, y_pred and sensitive_features (and kwargs)

File: api/METRICS.md (hunk 5 above)
Comment (riedgar-ms) [reply api/METRICS.md:52]: I have a severe allergy to undefined behaviour in programming languages. However, I'm not sure what would be considered the Pythonic thing to do.

File: api/METRICS.md (hunk 6 above)
Comment (riedgar-ms) [reply api/METRICS.md]: OK.

File: api/METRICS.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+## Proposal
+
+*Function signatures*
+
+```python
+group_summary(metric, y_true, y_pred, *, sensitive_features, **other_kwargs)
+# return the group summary for the provided `metric`, where `metric` has the signature
+# metric(y_true, y_pred, **other_kwargs)
+
+make_metric_group_summary(metric)
```
Comment (riedgar-ms) [inline api/METRICS.md:47]: OK, although for strict comparison to `make_scorer` it should be `make_metric_group_summariser()`

File: api/METRICS.md (hunk 5 above)
Comment (MiroDudik) [reply api/METRICS.md:52]: for start, let's just support scalars and throw exceptions otherwise.

File: api/METRICS.md (hunk 9 above)
Comment (romanlutz) [reply api/METRICS.md:47]: You mean `make_metric_group_summarizer()` ;-)

**Review by romanlutz — APPROVED** [review]

File: api/METRICS.md (hunk 9 above)
Comment (MiroDudik) [reply api/METRICS.md:47]: my rationale for `make_metric_group_summary` is that the result is the function to which we will refer to as `<metric>_group_summary` elsewhere in the API; that in turn makes sense since other derived metrics are called `<metric>_difference`, `<metric>_ratio` etc.

File: api/METRICS.md (hunk 5 above)
Comment (riedgar-ms) [reply api/METRICS.md:52]: Suits me

File: api/METRICS.md (hunk 9 above)
Comment (riedgar-ms) [reply api/METRICS.md:47]: OK

**Review by riedgar-ms — APPROVED** [review]

File: api/METRICS.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+|`mean_squared_error`| G | prob, reg | sklearn | - |
+|`r2_score`| G | reg | sklearn | - |
+
+## Dashboard questions
+
+1. Should we enable regression metrics for probabilistic classification?
+   * `mean_absolute_error`, `mean_squared_error`, `mean_squared_error(...,squared=False)`
+1. Should we introduce balanced error metrics for probabilistic classification?
+   * `balanced_mean_{squared,absolute}_error`, `balanced_log_loss`
+1. Do we keep `mean_prediction` and `mean_{over,under}prediction`?
```
Comment (rihorn2) [inline api/METRICS.md]: Unless we no longer need it in the dashboard, it will need to be available in python. We could special case these and not have them callable by the user if we want. (what about selection rate? the other ones like our differently normalized underprediction rate)?

**Review by rihorn2 — APPROVED** [review]
overall, this looks like it minimally impacts the dashboard, save for the few special cases listed below.

**Comment by adrinjalali** [comment]
I know y'all have a very good understanding of this proposal, but as somebody who's not as familiar with the API as you are, I have a hard time following the proposal. It'd be nice if you could have a motivation, explain the status quo, and explain what the proposed API does and how. It'd make it easier for the rest of us to follow :)

File: api/METRICS.md (hunk 10 above)
Comment (MiroDudik) [reply api/METRICS.md]: For the time being, I'll proceed with non-breaking changes. We will need to review metrics / dashboard integration in a separate PR.
